codec-http/src/main/java/io/netty/handler/codec/http/QueryStringEncoder.java

```diff
@@ -95,27 +116,54 @@ public String toString() {
         return uriBuilder.toString();
     }
 
-    private static void appendComponent(String s, String charset, StringBuilder sb) {
-        try {
-            s = URLEncoder.encode(s, charset);
-        } catch (UnsupportedEncodingException ignored) {
-            throw new UnsupportedCharsetException(charset);
-        }
-        // replace all '+' with "%20"
-        int idx = s.indexOf('+');
-        if (idx == -1) {
-            sb.append(s);
-            return;
-        }
-        sb.append(s, 0, idx).append("%20");
-        int size = s.length();
-        idx++;
-        for (; idx < size; idx++) {
-            char c = s.charAt(idx);
-            if (c != '+') {
-                sb.append(c);
+    /**
+     * encode the String as per RFC 3986, Section 2.
+     * <p>
+     * There is a little different between the JDK's encode method : {@link URLEncoder#encode(String, String)}.
+     * The JDK's encoder encode the space to {@code +} and this method directly encode the blank to {@code %20}
+     * beyond that , this method reuse the {@link #uriBuilder} in this class rather then create a new one,
+     * thus generates less garbage for the GC.
+     *
+     * @param s The String to encode
+     */
+    private void encodeComponent(String s) {
+        //allocate memory until needed
+        char[] buf = null;
+
+        for (int i = 0; i < s.length();) {
+            int c = s.charAt(i);
+            if (dontNeedEncoding.get(c)) {
+                uriBuilder.append((char) c);
+                i++;
             } else {
-                sb.append("%20");
+                int index = 0;
+                if (buf == null) {
+                    buf = new char[s.length() - i];
+                }
+
+                do {
+                    buf[index] = (char) c;
+                    index++;
+                    i++;
+                } while (i < s.length() && !dontNeedEncoding.get(c = s.charAt(i)));
+
+                byte[] bytes = new String(buf, 0, index).getBytes(charset);
+
```

Comment:
Why not use indexOf directly and so no create a byte array ? Also maybe use AsciiString

Reply:
The String created here contains the unencoded chars ,these chars may be multi-byte char so AsciiString can not be used here.

Reply:
And you mean use indexOf to get a single char and use CharsetEncoder rather then create a String?I test this approach before,the result was not good at that time.I can test again at tomorrow morning

Reply:
I see...